torch/csrc/jit/autodiff.cpp

```diff
@@ -34,38 +45,233 @@ static std::vector<Value*> gradientForNode(Node* node, ArrayRef<Value*> grad_val
   return fmap(sym_grads, [](const SymbolicVariable &v) { return v.value(); });
 }
 
-void differentiate(std::shared_ptr<Graph>& graph) {
-  JIT_ASSERT(graph->stage() == 0);
-  graph->advanceStage();
+// Before:
+//   - graph has only stage 0
+//   - grad_desc doesn't have any fields initialized
+// After:
+//   - graph has stage 0 and stage 1 that computes its vjp
+//   - grad_desc has df_input_vjps and df_output_vjps set
+//     (but df_input_vjps will be modified later as well)
+static value_map addReverseInline(Graph& graph, Gradient& grad_desc) {
+  JIT_ASSERT(graph.stage() == 0);
+  graph.advanceStage();
+
+  value_map grad_map; // x -> dx mapping
+  const auto get_grad = [&](Value* v) { return grad_map.at(v); };
+  const auto set_grad = [&](Value *x, Value *dx) {
+    if (Value * prev_grad = grad_map[x]) {
+      Value * new_grad = addAndPutAfter(prev_grad, dx, dx->node());
+      grad_map[x] = new_grad;
+    } else {
+      grad_map[x] = dx;
+    }
+  };
 
-  std::unordered_map<Value*, Value*> grad_map; // x -> dx mapping
-  const auto get_grad = [&](Value* v) { return grad_map[v]; };
-  for (auto output : graph->outputs())
-    grad_map[output] = graph->addInput()->setType(output->typeOption());
+  auto outputs = graph.outputs();
+  for (std::size_t i = 0, num_outputs = outputs.size(); i < num_outputs; ++i) {
+    Value * output = outputs[i];
+    Value * output_grad = graph.addInput()->setType(output->typeOption());
+    set_grad(output, output_grad);
+    grad_desc.df_input_vjps.push_back(i);
```

Comment:
So, this seems to imply that for i < num_outputs, grad_desc.df_input_vjps[i] == i, is that right? Will we ever reorder this?

Reply:
Yes, that's right and it will be like this until we add `requires_grad` (when `df_input_vjps` will be a subset of this interval).